src/ring_theory/rees_algebra.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+
+import ring_theory.finiteness
+import ring_theory.ideal.local_ring
+import ring_theory.nakayama
+
+/-!
+
+# Rees algebra
+
+The Rees algebra of an ideal `I` is the subalgebra `R[It]` of `R[X]` defined as `R[It] = ⨁ₙ Iⁿ tⁿ`.
+This is used to prove the Artin-Rees lemma, and will potentially enable us to calculate some
+blowup in the future.
+
+## Main definition
+
+- `rees_algebra` : The Rees algebra of an ideal `I`, defined as a subalgebra of `R[X]`.
+- `adjoin_monomial_eq_rees_algebra` : The Rees algebra is generated by the degree one elements.
+- `rees_algebra.fg` : The Rees algebra of a f.g. ideal is of finite type. In particular, this
+implies that the rees algebra over a noetherian ring is still noetherian.
+
+-/
+
+universes u v
+
+variables {R M : Type u} [comm_ring R] [add_comm_group M] [module R M] (I : ideal R)
+
+open polynomial
+open_locale polynomial big_operators
+
+/-- The Rees algebra of an ideal `I`, defined as the subalgebra of `R[X]` whose `i`-th coefficient
+falls in `I ^ n`. -/
+def rees_algebra : subalgebra R R[X] :=
+{ carrier := { f | ∀ i, f.coeff i ∈ I ^ i },
+  mul_mem' := λ f g hf hg i, begin
+    rw coeff_mul,
+    apply ideal.sum_mem,
+    rintros ⟨j, k⟩ e,
+    rw [← finset.nat.mem_antidiagonal.mp e, pow_add],
+    exact ideal.mul_mem_mul (hf j) (hg k)
+  end,
+  one_mem' := λ i, begin
+    rw coeff_one,
+    split_ifs,
+    { subst h, simp },
+    { simp }
+  end,
+  add_mem' := λ f g hf hg i, begin
+    rw coeff_add,
+    exact ideal.add_mem _ (hf i) (hg i)
+  end,
+  zero_mem' := λ i, ideal.zero_mem _,
+  algebra_map_mem' := λ r i, begin
+    rw [algebra_map_apply, coeff_C],
+    split_ifs,
+    { subst h, simp },
+    { simp }
+  end }
+
+lemma mem_rees_algebra_iff (f : R[X]) :
+  f ∈ rees_algebra I ↔ ∀ i, f.coeff i ∈ I ^ i := iff.rfl
```

Comment:
`rees_algebra.monomial_mem` and `adjoin_monomial_eq_rees_algebra` seem to require a slightly too difficult proof. Does the situation improve if we strengthen this to:
```suggestion
lemma mem_rees_algebra_iff (f : R[X]) :
  f ∈ rees_algebra I ↔ ∀ i ∈ f.support, f.coeff i ∈ I ^ i := _
```